capirca/lib/fortigate.py

```diff
@@ -0,0 +1,330 @@
+      return
+    type(self).FW_ADDRESSES.extend(['\tedit %s' % addr,
+                                    '\t\tset subnet %s' % addr,
+                                    '\tnext'])
+    type(self)._FW_DUP_CHECK.add(addr)
+
+  def _add_service_to_fw_services(self, protocol, service):
+    if service in type(self)._FW_DUP_CHECK:
+      return
+
+    type(self).FW_SERVICES.extend(['\tedit %s' % service,
+                                   '\t\tset protocol TCP/UDP',
+                                   '\t\tset %s-portrange %s' % (protocol.lower(), service),
+                                   '\tnext'])
+
+    type(self)._FW_DUP_CHECK.add(service)
+
+  def _generate_address_names(self, *addresses):
+    for group in addresses:
+      for addr in group:
+        if addr and not isinstance(addr, nacaddr.IPv6):
+          self._add_address_to_fw_addresses(addr.with_prefixlen)
+
+  def __str__(self):
+    lines = []
+
+    self._generate_address_names(self._term.destination_address, self._term.source_address)
+    # lines.extend(self.firewall_addresses)
+
+    dest_addresses = self._get_addresses_name(self._term.destination_address)
+    src_addresses = self._get_addresses_name(self._term.source_address)
+    services = self._get_services_string(self._term.protocol, self._term.destination_port)
+
+    lines.append('\t\tset comments %s' % self._term.name)
+    lines.append('\t\tset srcintf %s' % (self._term.source_interface or 'any'))
+    lines.append('\t\tset dstintf %s' % (self._term.destination_interface or 'any'))
+    lines.append('\t\tset dstaddr %s' % dest_addresses)
+    lines.append('\t\tset srcaddr %s' % src_addresses)
+    lines.append('\t\tset action %s' % _ACTION_TABLE.get(self._term.action[0]))
+    lines.append('\t\tset service %s' % services)
+    lines.append('\t\tset schedule always')
+    if self._term.logging:
+      lines.append('\t\tset logtraffic all')
+
+    return '\n'.join(lines)
+
+
+class Fortigate(aclgenerator.ACLGenerator):
+  """A cisco policy object."""
+
+  _PLATFORM = 'fortigate'
+  _DEFAULT_PROTOCOL = 'ALL'
+  SUFFIX = '.fcl'
+  # Protocols should be emitted as numbers.
+  _PROTO_INT = True
+  _TERM_REMARK = True
+
+  def _BuildTokens(self):
+    """Build supported tokens for platform.
+
+    Returns:
+      tuple containing both supported tokens and sub tokens
+    """
+    supported_tokens, supported_sub_tokens = super(Fortigate, self)._BuildTokens()
+
+    supported_tokens |= {'source_interface',
+                         'destination_interface',
+                         'logging'}
+
+    supported_sub_tokens.update({'option': {'from_id'},
+                                 # Warning, some of these are mapped
+                                 # differently. See _ACTION_TABLE
+                                 'action': {'accept', 'deny', 'reject',
+                                            'reject-with-tcp-rst'}})
+    return supported_tokens, supported_sub_tokens
+
+  def _TranslatePolicy(self, pol, exp_info):
```

Comment:
Docstring here please.

Reply:
docstrings should contain the arguments returns and other sections.
Take a look at some of the docstrings in junipersrx
https://github.com/google/capirca/blob/master/capirca/lib/junipersrx.py#L228
There are sections for the arguments, returns, methods, and attributes where applicable.

Please fix this here and elsewhere.